### ISSUE TITLE
pe.section_exists(name)

### ISSUE THREAD

**Issue by dzzie** [VirusTotal/yara#1238]
I think pe.section_exists(name) would be a useful addition and make for cleaner sigs over having to write an inline for loop in the script to see if it exists. 

'''C
//old: 
for any i in (0 .. pe.number_of_sections-1):(pe.sections[i].name == ".symtab")

//new: 
pe.section_exists(".symtab")

declare_function("section_exists", "s", "i", section_name_exists);

define_function(section_name_exists)
{
	YR_OBJECT* module = module();

	char* name = string_argument(1);

	int64_t n = get_integer(module, "number_of_sections");
	int i;

	if (is_undefined(module, "number_of_sections"))
		return_integer(0);

	for (i = 0; i < yr_min(n, MAX_PE_SECTIONS); i++)
	{
		SIZED_STRING* sect = get_string(module, "sections[%i].name", i);

		if (sect != NULL && strcmp(name, sect->c_string) == 0)
			return_integer(1);
	}

	return_integer(0);
}
```

**Comment by dzzie** [comment]
alternatively if section_index(name) returned -1 instead of undefined for a non existent section this would suffice as well.

**Comment by wxsBSD** [comment]
If you want to see if a section exists you can just do `pe.section_index(".foo") >= 0`. If the section name does not exist you end up with `UNDEFINED >= 0` which is always false. However, your comment about a `pe.section_index(/foo/)` is valid IMO. Being able to search for section names by regex is useful for sure.

**Comment by dzzie** [comment]
I also want to be able to check if the section does not exist 

not pe.section_exists(".symtab") //works
not for any i in (0 .. pe.number_of_sections-1):(pe.sections[i].name == ".symtab") //works
not (pe.section_index(".symtab") >= 0) // does not work

The only way I have found to test for UNDEFINED a custom extension such as pe.isdef("rich_signature.offset") or to wrap it in its own rule

```c
private rule richHeaderExists { condition: pe.rich_signature.offset  }
rule noRichHeader{  condition: not richHeaderExists }
```

**Comment by wxsBSD** [comment]
Ah, I see now. Thank you for the clarification. I wasn't accurate earlier when I said `UNDEFINED >= 0` will always be false. It's actually it will always be `UNDEFINED.` Making `section_index()` return -1 if the name isn't found makes sense, but I'm not sure the implications here. Should we make all the similar functions operate like that?

**Comment by dzzie** [comment]
Yeah with so many rules already written changing section_index to switch from UNDEFINED to -1 would have to be very carefully considered by those with magic greater than mine.

Having an UNDEFINED value that you can not test for is seems to be the root my problems in several cases so far.

If there was a valid == UNDEFINED or the pe.isdef(string_to_Eval) method available, these seem like a good generic way to possibly go forward without headache of a possible conflict or breaking change.

The isdef method I implemented for my tool is really simple and covers many cases:

```C
//declare_function("isdef", "s", "i", isdef); 
//ex: pe.dbg(pe.isdef("rich_signature.offset"))
//    pe.dbg(pe.isdef("number_of_signatures"))
define_function(isdef) 
{
	YR_OBJECT* module = module();
	char* txt = string_argument(1);
	if (yr_object_has_undefined_value(module, txt)){
		return_integer(0);
	}else{
		return_integer(1);
	}
}
```

In this case however it still could not cover the section_index(name) scenario so I had to resort to a section_exists(name) name. (While the not for loop works its a lot to type and get the syntax just right)

**Comment by plusvic** [comment]
Actually, as part of the fix for issue #1180, I'm changing the behavior for comparison operators. When some operand is `UNDEFINED`, the result is `false` instead of `UNDEFINED`, so not `not (pe.section_index(".symtab") >= 0)` will work in YARA 4.0. 

For arithmetic and bitwise operations the result is still `UNDEFINED` if some operand is `UNDEFINED`

**Comment by strictlymike** [comment]
If this means that I will be able to compare the result with a value that can be expressed in Yara (e.g. `false`) then I'm for it. I would also be ok with it if I were able to express `UNDEFINED` and compare the current return value with that.

**Comment by strictlymike** [comment]
For versions that don't include a fix for this, I think the below is a usable work-around:

```
        for all i in (0 .. (pe.number_of_sections - 1)):
        (
            pe.sections[i].name != ".text"
        )
```
